pallets/corporate-actions/src/ballot.rs

```diff
@@ -0,0 +1,544 @@
+        ///    exceeds `origin`'s DID's voting power.
+        #[weight = 950_000_000]
+        pub fn vote(origin, ca_id: CAId, votes: Vec<(T::Balance, u16)>) {
+            let did = <Identity<T>>::ensure_perms(origin)?;
+
+            // Ensure ballot has started but not ended, i.e. `start <= now <= end`.
+            let range = Self::ensure_ballot_exists(ca_id)?;
+            let now = <Checkpoint<T>>::now_unix();
+            ensure!(range.start <= now, Error::<T>::VotingNotStarted);
+            ensure!(now <= range.end, Error::<T>::VotingAlreadyEnded);
+
+            // Ensure we have balances provided for each choice.
+            let choices_count = MotionNumChoices::get(ca_id);
+            let total_choices = choices_count.iter().copied().map(|c| c as usize).sum::<usize>();
+            ensure!(votes.len() == total_choices, Error::<T>::WrongVoteCount);
+
+            // Divide `votes` into motions.
+            let motions = choices_count
+                .iter()
+                .map(|c| *c as usize)
+                .scan(0, |start, count| Some(&votes[mem::replace(start, *start + count)..count]));
+
+            if RCV::get(ca_id) {
+                // RCV is enabled.
+                // Ensure that all fallback choices point to some choice in the same motion.
+                motions
+                    .clone()
+                    .all(|votes| {
+                        let count = votes.len();
+                        votes
+                            .iter()
+                            .map(|vote| vote.1 as usize)
+                            .enumerate()
+                            .all(|(idx, fb)| idx != fb && fb < count)
+                    })
+                    .then_some(())
+                    .ok_or(Error::<T>::NoSuchRCVFallback)?;
+            } else {
+                // It's not. Make sure its also not used.
+                votes
+                    .iter()
+                    .all(|(_, fb)| *fb == 0)
+                    .then_some(())
+                    .ok_or(Error::<T>::RCVNotAllowed)?;
+            }
+
+            // Ensure that `did` is targeted by this ballot.
+            let ca = <CA<T>>::ensure_ca_exists(ca_id)?;
+            ensure!(ca.targets.targets(&did), Error::<T>::NotTargetedByCA);
+
+            // Extract `did`'s balance at the record date.
+            // Record date has passed by definition.
+            let ticker = ca_id.ticker;
+            let cp_id = match ca.record_date.unwrap().checkpoint {
+                CACheckpoint::Existing(id) => Some(id),
+                CACheckpoint::Scheduled(id) => <Checkpoint<T>>::schedule_points((ticker, id)).pop(),
+            };
+            let available_power = match cp_id {
+                // CP exists, use it.
+                Some(cp_id) => <Asset<T>>::get_balance_at(ticker, did, cp_id),
+                // No transfer yet. Use current balance instead.
+                None => <Asset<T>>::balance_of(ticker, did),
```

Comment:
So - this means the checkpoint is the most recent checkpoint, and there have been no asset transfers since the checkpoint?

Reply:
Indeed no transfers have happened since record date, so the CP, although due, was never created, so the current balance / supply is used instead. Clarifying the comment per discussion on a meeting.